lib/std/compress/deflate/decompressor.zig

```diff
@@ -0,0 +1,997 @@
+                    // Longer codes should have been
+                    // associated with a link table above.
+                    unreachable; // "impossible: not an indirect chunk"
+                }
+                var value = self.chunks[j] >> huffman_value_shift;
+                var link_tab = self.links[value];
+                reverse >>= huffman_chunk_bits;
+                var off = reverse;
+                while (off < link_tab.len) : (off += @as(u32, 1) << @intCast(u5, n - huffman_chunk_bits)) {
+                    if (sanity and link_tab[off] != 0) {
+                        unreachable; // "impossible: overwriting existing chunk"
+                    }
+                    link_tab[off] = chunk;
+                }
+            }
+        }
+
+        if (sanity) {
+            // Above we've sanity checked that we never overwrote
+            // an existing entry. Here we additionally check that
+            // we filled the tables completely.
+            for (self.chunks) |chunk, i| {
+                if (chunk == 0) {
+                    // As an exception, in the degenerate
+                    // single-code case, we allow odd
+                    // chunks to be missing.
+                    if (code == 1 and i % 2 == 1) {
+                        continue;
+                    }
+                    unreachable; // "impossible: missing chunk"
+                }
+            }
+            for (self.links) |link_tab| {
+                for (link_tab) |chunk| {
+                    if (chunk == 0) {
+                        unreachable; // "impossible: missing chunk"
+                    }
+                }
+            }
+        }
+
+        return true;
+    }
+
+    pub fn deinit(self: *Self) void {
+        if (self.initialized and self.links.len > 0) {
+            for (self.sub_links.items) |off| {
+                self.allocator.free(self.links[off]);
+            }
+            self.allocator.free(self.links);
+            self.sub_links.deinit();
+        }
+    }
+};
+
+var fixed_huffman_decoder: ?HuffmanDecoder = null;
+
+fn fixedHuffmanDecoderInit(allocator: Allocator) !HuffmanDecoder {
+    if (fixed_huffman_decoder != null) {
+        return fixed_huffman_decoder.?;
+    }
+
+    // These come from the RFC section 3.2.6.
+    var bits: [288]u32 = undefined;
+    var i: u32 = 0;
+    while (i < 144) : (i += 1) {
+        bits[i] = 8;
+    }
+    while (i < 256) : (i += 1) {
+        bits[i] = 9;
+    }
+    while (i < 280) : (i += 1) {
+        bits[i] = 7;
+    }
+    while (i < 288) : (i += 1) {
+        bits[i] = 8;
+    }
+
+    fixed_huffman_decoder = HuffmanDecoder{};
+    _ = try fixed_huffman_decoder.?.init(allocator, &bits);
+    return fixed_huffman_decoder.?;
+}
+
+const DecompressorState = enum {
+    Init,
+    Dict,
+};
+
+pub fn decompressor(allocator: Allocator, reader: anytype, dictionary: ?[]const u8) !Decompressor(@TypeOf(reader)) {
```

Comment:
The fact that this requires an allocator is an immediate indicator that it is *not* in any way obvious that "any puff.c based implementation should be replaced."

Reply:
The fact that this requires an allocator may mean to you that the new implementation can't replace the current one. Which I agree with, if you don't want an allocator in there. But my original statement stands. "any puff.c based implementation should be replaced." if you want speed at least.

Reply:
This decompressor uses 2 levels of linked lookup tables (chunks[] & links[]/sub_chunks) to decode a Huffman tree.
1 tree uses stack memory (fixed huffman) (only uses chunks[])
2 trees uses heap memory (dynamic huffman) (uses chunks[] and links[]/sub_chunks)
The size of the tables depends on the series of code lengths describing each tree.
The first table size depends on the number of chunks which at max can be 512.
512 * []u16 = 8_192 bytes
The second table size depends on the max code length describing the tree and the size in bits of a chunk.
512 * (1<<(max code length - size in bits of chunk))
512 * (1<<(16 - 9))
512 * 128
[65536]u16 = 131_072 bytes
So 262_144 bytes for the 2 dynamic Huffman trees.
The dictionary decoder copies the [32768]u8 window which adds 32768 bytes
bits is [286+30]u32 = 1264 bytes
codebits is [19]u32 = 76 bytes
The maximum allocated heap memory should be 304_444 bytes.

With a recent understanding that a FixedBufferAllocator won't release unaligned memory,I managed to make a prototype work with a FixedBuffferAllocator initialized with a buffer of size 304444 bytes. (I replaced sub_chunks ArrayList by a [128]u16 array). Unfortunately, I can't get it to work on wasm. I'm lacking knowledge about memory alignment. Maximum resident size was ~700kb which is around the same as the GeneralPurposeAllocator (with an ArrayList for sub_chunks).

I added a note in the triple slash documentation of `decompressor()`

Reply:
Thank you for this analysis :+1: